solver/src/solver/paraswap_solver/api.rs

```diff
@@ -45,24 +46,76 @@ impl ParaswapApi for DefaultParaswapApi {
     async fn transaction(
         &self,
         query: TransactionBuilderQuery,
-    ) -> Result<TransactionBuilderResponse> {
-        tracing::debug!(
-            "Querying Paraswap API (transaction) with {}",
-            serde_json::to_string(&query).unwrap()
-        );
-        let text = query
-            .into_request(&self.client)
-            .send()
-            .await
-            .context("TransactionBuilderQuery failed")?
-            .text()
-            .await?;
-        tracing::debug!("Response from Paraswap API (transaction): {}", text);
+    ) -> Result<TransactionBuilderResponse, ParaswapResponseError> {
+        let query_str = serde_json::to_string(&query).unwrap();
+        let response_result = query.into_request(&self.client).send().await;
+        match response_result {
+            Ok(response) => match response.text().await {
+                Ok(response_text) => parse_paraswap_response_text(&response_text, &query_str),
+                Err(text_fetch_err) => Err(ParaswapResponseError::TextFetch(text_fetch_err)),
+            },
+            Err(send_err) => Err(ParaswapResponseError::Send(send_err)),
+        }
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum RawResponse {
+    ResponseOk(TransactionBuilderResponse),
+    ResponseErr { error: String },
+}
+
+#[derive(Error, Debug)]
+pub enum ParaswapResponseError {
+    // Represents a failure with transaction builder query
+    #[error("ERROR_BUILDING_TRANSACTION from query {0}")]
+    BuildingTransaction(String),
+
+    // Occurs when the price changes between the time the price was queried and this request
+    #[error("Suspected Rate Change - Please Retry!")]
+    PriceChange,
+
+    // Connectivity or non-response error
+    #[error("Failed on send")]
+    Send(reqwest::Error),
+
+    // Recovered Response but failed on async call of response.text()
+    #[error(transparent)]
+    TextFetch(reqwest::Error),
+
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
```

Comment:
If we are using thiserror do distinguish all the other errors anyway then let's also do that here

```suggestion
    #[error("unknown paraswap error: {0}")]
    UnknownParaswapError(String)

    #[error(transparent)]
   DeserializeError(#[from] serde_json::Error)
```